client/src/hooks/board/useMenuActions.ts

```diff
@@ -31,142 +35,201 @@ const getProperties = (color: string, size: number): CanvasActionProperties => (
-  },
-  [MenuActions.ADD_IMAGE_URL]: ({ setCanvasMode }) => {
-    if (!setCanvasMode) {
-      return;
-    }
-    setCanvasMode(CanvasMode.SELECTION);
-  },
-  [MenuActions.ADD_IMAGE_DISK]: ({ setCanvasMode }) => {
-    if (!setCanvasMode) {
-      return;
-    }
-    setCanvasMode(CanvasMode.SELECTION);
-  },
-  [MenuActions.LOAD_IMAGES_JSON]: ({ setCanvasMode }) => {
-    if (!setCanvasMode) {
-      return;
-    }
-    setCanvasMode(CanvasMode.SELECTION);
-  },
-  [MenuActions.CHANGE_COLOR]: () => {
-    // No-op: handled by color picker UI
-    throw new Error("Change () => performAction(MenuActions.CHANGE_COLOR) to () => {} in MenuDataContext - no effect");
-  },
-  [MenuActions.CHANGE_SIZE]: () => {
-    // No-op: handled by size picker UI
-    throw new Error("Change () => performAction(MenuActions.CHANGE_SIZE) to () => {} in MenuDataContext - no effect");
-  },
-  [MenuActions.EXPORT_PDF]: ({ canvas }) => {
-    if (!canvas) {
-      return;
-    }
-    exportToPDF(canvas);
-  },
-  [MenuActions.UNDO]: () => {
-    // No-op: handled directly in MenuDataContext
-    throw new Error(
-      "Change () => performAction(MenuActions.UNDO) to undo() - it should be called directly in MenuDataContext"
-    );
-  },
-  [MenuActions.REDO]: () => {
-    // No-op: handled directly in MenuDataContext
-    throw new Error(
-      "Change () => performAction(MenuActions.REDO) to redo() - it should be called directly in MenuDataContext"
-    );
-  },
-  [MenuActions.SAVE_CANVAS]: ({ canvas }) => {
-    if (!canvas) {
-      return;
-    }
-    handleSave(canvas);
-  },
-  [MenuActions.SAVE_IMAGES]: ({ canvas }) => {
-    if (!canvas) {
-      return;
-    }
-    saveImagesToLocalFile(canvas);
-  },
-};
-
 export const useMenuActions = () => {
   const {
     state: { canvas, color, size },
     setCanvasMode,
   } = useCanvas();
-  const { saveState } = useUndoRedo();
+  const { saveCommand } = useUndoRedo();
+
+  const actionHandlers: Record<MenuActions | CanvasMode, CanvasActionHandler> = useMemo(() => {
+    const actionHandlers: Record<MenuActions | CanvasMode, CanvasActionHandler> = {
+      [CanvasMode.SELECTION]: ({ setCanvasMode }) => {
+        if (!setCanvasMode) {
+          return;
+        }
+        setSelectionMode(setCanvasMode);
+      },
+      [CanvasMode.SIMPLE_DRAWING]: ({ canvas, color, size, setCanvasMode }) => {
+        if (!canvas || !setCanvasMode) {
+          return;
+        }
+        setDrawingMode(canvas, color as string, size as number, setCanvasMode);
+      },
+      [CanvasMode.ERASER]: ({ canvas, size, setCanvasMode }) => {
+        if (!canvas || !setCanvasMode) {
+          return;
+        }
+        setEraserMode(canvas, size as number, setCanvasMode);
+      },
+      [MenuActions.ADD_LINE]: ({ canvas, properties, setCanvasMode }) => {
+        if (!canvas || !properties || !setCanvasMode) {
+          return;
+        }
+        const addedLine = addLine(canvas, properties);
```

Comment:
It seems like there's a lot of repeated code when adding new objects. I assume you weren't sure how to refactor this elegantly, since the logic for GROUP_SELECTED and the code below it differs?

Reply:
To be honest I haven't even bothered with eliminating code repetition, although I should have 😅
I will look into it